[PyTorch] Move prim string ops to JIT op registry

larryliu0820 opened the pull request:
Summary: This PR migrates prim string ops to be registered into JIT op registry instead of dispatcher. Since the implementation of these ops are backend agnostic, there's no need to go through dispatcher. Relying on `test_jit_string.py` to verify the correctness of these ops. I'm also adding tests to make sure all the operators are covered.

Test Plan: Rely on `test_jit_string.py`.

Differential Revision: D33351638

bdhirsh commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -1494,524 +1494,660 @@ int64_t stringFindImpl(
+
+RegisterOperators regStrOps(createOperators(
+    stringOpGenArgs,
+    sizeof(stringOpGenArgs) / sizeof(stringOpGenArgs[0])));
```
Are you doing the `sizeof()` calls to pass in the number of elements of the array to `createOperators()`?

Maybe make `stringOpGenArgs` a `std::array` instead, that way you can just call `.size()` instead.

larryliu0820 replied:
I was following the example here: https://fburl.com/tz3vkorc. I can try `std::array` instead.

bdhirsh reviewed (changes requested):
LGTM other than the minor comments! Maybe @iseeyuan should also take a look.

iseeyuan reviewed (approved):
LGTM!

iseeyuan reviewed (commented):
@bdhirsh and @larryliu0820 , I saw catchall in other op registrations (majorly quantization related). What's the plan of those operators? Could we also move them to JIT registration? My assumption is that the overhead would be smaller because we don't need to go through the dispatch key. Also would that be easier for static dispatch?

bdhirsh commented:
> @bdhirsh and @larryliu0820 , I saw catchall in other op registrations (majorly quantization related). What's the plan of those operators? Could we also move them to JIT registration? My assumption is that the overhead would be smaller because we don't need to go through the dispatch key. Also would that be easier for static dispatch?

Maybe someone who wrote those ops can add more clarity. The way that I understand it is:
- if the op is *only* used by mobile / torchscript, then it seems reasonable to register directly to the jit registry
- if the op is meant to be useable in eager mode, then we want it registered to the dispatcher.

larryliu0820 committed:
[PyTorch] Move prim string ops to JIT op registry (#70501)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/70501

This PR migrates prim string ops to be registered into JIT op registry instead of dispatcher. Since the implementation of these ops are backend agnostic, there's no need to go through dispatcher. Relying on `test_jit_string.py` to verify the correctness of these ops. I'm also adding tests to make sure all the operators are covered.

Test Plan: Rely on `test_jit_string.py`.

Reviewed By: iseeyuan

Differential Revision: D33351638

fbshipit-source-id: 1df8a931a7a03b96ad8502ca12976e9033198ebd
